### PR TITLE
Use dynamic list of document formats

### PR DESCRIPTION
**Pull request by vanitabarrett**
Trello card: https://trello.com/c/dwMWGd0l/69-get-the-list-of-available-formats-document-types

This PR removes a hardcoded list of document formats and replaces it by reading in [Rummager's list of formats](https://github.com/alphagov/rummager/blob/master/config/govuk_index/migrated_formats.yaml).

We don't want to hardcode a list of document formats into the prototype, as this would mean manually maintaining the list when formats are removed/added. We've chosen to use Rummager's list of 'migrated formats.yml' as this closely matches the formats that are being returned in our Rummager queries.

**Investigation into lists of document types/formats/supergroups [here](https://docs.google.com/document/d/1ohOObFMPVQwKxdd7rEWeydKDCgit3vP6fL3htbJ9iVg/edit#)**

**Note: there are some formats which are not listed in this yml file, so we are having to also hardcode those into our prototype.**

### CODE REVIEW

File: app/controllers/segment_constraints/document_type_constraint.rb [hunk 1]
```diff
@@ -1,6 +1,24 @@
+require "open-uri"
+
 class DocumentTypeConstraint
+  DOCUMENT_TYPE_YAML = "https://raw.githubusercontent.com/alphagov/rummager/master/config/govuk_index/migrated_formats.yaml".freeze
+
+  # Rummager's migrated formats yml doesn't contain all formats, so we need to hardcode these into our prototype
+  HARDCODED_FORMATS = ["residential_property_tribunal_decision", "task_list", "calculator", "field_of_operation", "html_publication", "imported", "ministerial_role", "organisation", "person", "policy_area", "services_and_information", "take_part", "topical_event", "topical_event_about_page", "working_group", "world_location", "worldwide_organisation", "corporate_information_page", "closed_consultation", "consultation_outcome", "esi_fund", "external_content", "national", "news_article", "official", "open_consultation", "statistics_announcement", "topic", "world_location_news_article"]
```
Comment (leenagupte) [inline app/controllers/segment_constraints/document_type_constraint.rb]: Something I've been bitten by in the past is that [govuk-lint](https://github.com/alphagov/govuk-lint/blob/master/configs/rubocop/gds-ruby-styleguide.yml#L308) makes you change arrays of single words to a word array. 

We're not using the linter here, so it's not a problem.

File: app/controllers/segment_constraints/document_type_constraint.rb [hunk 2]
```diff
@@ -1,6 +1,24 @@
...
-    @doc_types = YAML.load_file(Rails.root.join("lib", "data", "formats.yml"))
+    @doc_types = load_document_types
+  end
+
+  def load_document_types
+    document_type_yaml_file = open(DOCUMENT_TYPE_YAML) {|f| f.read}
+    document_type_yaml_data = YAML::load(document_type_yaml_file)
+
+    document_type_yaml_data.each do |title, doc_formats|
+      doc_formats.reject! { |doc_format| doc_format.is_a? Hash }
```
Comment (leenagupte) [inline app/controllers/segment_constraints/document_type_constraint.rb]: `reject!` has some interesting behaviour, in that it returns `nil`[1] if it doesn't have anything to reject. That doesn't seem to affected the results though.

[1] https://apidock.com/ruby/Array/delete_if#1018-NOT-Equivalent-to-Array-reject-

File: app/controllers/segment_constraints/document_type_constraint.rb [hunk 3]
```diff
@@ -1,6 +1,24 @@
...
+  HARDCODED_FORMATS = ["residential_property_tribunal_decision", "task_list", "calculator", "field_of_operation", "html_publication", "imported", "ministerial_role", "organisation", "person", "policy_area", "services_and_information", "take_part", "topical_event", "topical_event_about_page", "working_group", "world_location", "worldwide_organisation", "corporate_information_page", "closed_consultation", "consultation_outcome", "esi_fund", "external_content", "national", "news_article", "official", "open_consultation", "statistics_announcement", "topic", "world_location_news_article"]
+
   def initialize
-    @doc_types = YAML.load_file(Rails.root.join("lib", "data", "formats.yml"))
+    @doc_types = load_document_types
+  end
+
+  def load_document_types
+    document_type_yaml_file = open(DOCUMENT_TYPE_YAML) {|f| f.read}
+    document_type_yaml_data = YAML::load(document_type_yaml_file)
```
Comment (leenagupte) [inline app/controllers/segment_constraints/document_type_constraint.rb]: Minor thing. As we're getting the YAML file from an external site, I'd move the reading and loading of the YAML file to the `Services` module.

Comment (vanitabarrett) [reply app/controllers/segment_constraints/document_type_constraint.rb]: @leenagupte Good point 👍

File: app/controllers/segment_constraints/document_type_constraint.rb [hunk 4]
```diff
@@ -1,6 +1,24 @@
...
+
+    document_type_yaml_data.each do |title, doc_formats|
```
Comment (leenagupte) [inline app/controllers/segment_constraints/document_type_constraint.rb]: Like Python, Ruby lets you use `_` for unused variables. So you could write this line as:
`document_type_yaml_data.each do |_, doc_formats|`

**Review by leenagupte — APPROVED** [review]
One or two very minor points, and a couple of stray observations.

But LGTM! 👍

**Pull request merged**